alibi_detect/od/_pca.py

```diff
@@ -0,0 +1,184 @@
+from typing import Union, Optional, Callable, Dict, Any
+from typing import TYPE_CHECKING
+
+import numpy as np
+
+from alibi_detect.utils._types import Literal
+from alibi_detect.base import outlier_prediction_dict
+from alibi_detect.base import BaseDetector, ThresholdMixin, FitMixin
+from alibi_detect.od.pytorch import KernelPCATorch, LinearPCATorch
+from alibi_detect.utils.frameworks import BackendValidator
+from alibi_detect.version import __version__
+from alibi_detect.exceptions import _catch_error as catch_error
+
+
+if TYPE_CHECKING:
+    import torch
+
+
+backends = {
+    'pytorch': (KernelPCATorch, LinearPCATorch)
+}
+
+
+class PCA(BaseDetector, ThresholdMixin, FitMixin):
+    def __init__(
+        self,
+        n_components: int,
+        kernel: Optional[Callable] = None,
+        backend: Literal['pytorch'] = 'pytorch',
+        device: Optional[Union[Literal['cuda', 'gpu', 'cpu'], 'torch.device']] = None,
+    ) -> None:
+        """Principal Component Analysis (PCA) outlier detector.
+
+        The detector is based on the Principal Component Analysis (PCA) algorithm. There are two variants of PCA:
+        linear PCA and kernel PCA. Linear PCA computes the eigenvectors of the covariance matrix of the data. Kernel
+        PCA computes the eigenvectors of the kernel matrix of the data. In each case, we choose the smallest
+        `n_components` eigenvectors. We do this as they correspond to the invariant directions of the data. i.e the
+        directions along which the data is least spread out. Thus a point that deviates along these dimensions is more
+        likely to be an outlier.
+
+        When scoring a test instance we project it onto the eigenvectors and compute its score using the L2 norm. If
+        a threshold is fitted we use this to determine whether the instance is an outlier or not.
+
+        Parameters
+        ----------
+        n_components:
+            The number of dimensions in the principle subspace. For linear pca should have
+            ``1 <= n_components < dim(data)``. For kernel pca should have ``1 <= n_components < len(data)``.
```

Comment:
Since `n_components` is pretty fundamental to the algo, do we want to add a check and clear error message for the  `n_components >= dim(data)` case?